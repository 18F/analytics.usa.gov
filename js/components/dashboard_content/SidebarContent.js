import React from "react";
import PropTypes from "prop-types";
import { Tab, Tabs, TabList, TabPanel } from "react-tabs";

import TopDownloads from "./TopDownloads";
import TopPagesRealtime from "./TopPagesRealtime";
import TopPagesHistorical from "./TopPagesHistorical";
import TopVideoPlays from "./TopVideoPlays";

/**
 * Contains charts and other data visualizations for the top pages, top
 * downloads, and top video plays section of the site. This component is mainly
 * laying out the structure for the section, including tabs/panels, and passes
 * props necessary for getting data and displaying visualizations to child
 * components. The top pages chart, top downloads chart, and top video plays
 * displayed correspond to the time interval selected in the tabs.
 *
 * Note that the 30 minutes time interval is how often Google Analytics 4
 * updates realtime reports. The dimensions/metrics needed to report on file
 * downloads are not available in the realtime reporting API, so for the 30
 * minute time interval, show top downloads and top video plays for the past day
 * (the shortest time interval allowed for the non-realtime reporting API.)
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {string} props.agency the display name for the current agency.
 * @returns {import('react').ReactElement} The rendered element
 */
function SidebarContent({ dataHrefBase, agency }) {
  const numberOfTopPagesToDisplay = 30;
  const numberOfTopDownloadsToDisplay = 10;
  const numberOfTopVideoPlaysToDisplay = 10;

  return (
    <section className="sidebar-content">
      <div className="sidebar-content__headline">
        <h2>Top {numberOfTopPagesToDisplay} Web Pages and App Screens</h2>
      </div>
      <Tabs>
        <TabList className="usa-button-group usa-button-group--segmented">
          <Tab className="usa-button-group__item">
            <button type="button" className="usa-button">
              30 mins
            </button>
          </Tab>
          <Tab className="usa-button-group__item">
            <button type="button" className="usa-button">
              7 days
            </button>
          </Tab>
          <Tab className="usa-button-group__item">
            <button type="button" className="usa-button">
              30 days
            </button>
          </Tab>
        </TabList>

        <TabPanel>
          <section className="top-pages">
            <p>
              <strong>Users</strong> on a{" "}
              <strong>single, specific page or app screen</strong> in the last
              30 minutes. Hostnames are not currently reported in real-time, so
              only page title and screen name information is available.
            </p>
            <TopPagesRealtime
              dataHrefBase={dataHrefBase}
              reportFileName="top-pages-realtime.json"
              numberOfListingsToDisplay={numberOfTopPagesToDisplay}
              refreshSeconds={15}
            />
          </section>
          <section className="top-downloads">
            <div className="top-downloads__headline">
              <h3>Top {numberOfTopDownloadsToDisplay} Downloads Yesterday</h3>
            </div>
            <h4>
              <em>Total file downloads yesterday on {agency} hostnames.</em>
            </h4>
            <TopDownloads
              dataHrefBase={dataHrefBase}
              reportFileName="top-downloads-yesterday.json"
              numberOfListingsToDisplay={numberOfTopDownloadsToDisplay}
            />
          </section>
          <section className="top-video-plays">
            <div className="top-video-plays__headline">
              <h3>
                Top {numberOfTopVideoPlaysToDisplay} Video Plays Yesterday
              </h3>
            </div>
            <h4>
              <em>Total videos played yesterday on {agency} hostnames.</em>
            </h4>
            <TopVideoPlays
              dataHrefBase={dataHrefBase}
              reportFileName="top-video-plays-yesterday.json"
              numberOfListingsToDisplay={numberOfTopDownloadsToDisplay}
            />
          </section>
        </TabPanel>
        <TabPanel>
          <section className="top-pages">
            <p>
              Sessions over the last week on <strong>hostnames</strong>,
              including traffic to all web pages and app screens within that
              hostname.
            </p>
            <TopPagesHistorical
              dataHrefBase={dataHrefBase}
              reportFileName="top-domains-7-days.json"
              numberOfListingsToDisplay={numberOfTopPagesToDisplay}
            />
          </section>
          <section className="top-downloads">
            <div className="top-downloads__headline">
              <h3>Top {numberOfTopDownloadsToDisplay} Downloads Last 7 Days</h3>
            </div>
            <h4>
              <em>
                Total file downloads over the last week on {agency} hostnames.
              </em>
            </h4>
            <TopDownloads
              dataHrefBase={dataHrefBase}
              reportFileName="top-downloads-7-days.json"
              numberOfListingsToDisplay={numberOfTopDownloadsToDisplay}
            />
          </section>
          <section className="top-video-plays">
            <div className="top-video-plays__headline">
              <h3>
                Top {numberOfTopVideoPlaysToDisplay} Video Plays Last 7 Days
              </h3>
            </div>
            <h4>
              <em>
                Total videos played over the last week on {agency} hostnames.
              </em>
            </h4>
            <TopVideoPlays
              dataHrefBase={dataHrefBase}
              reportFileName="top-video-plays-7-days.json"
              numberOfListingsToDisplay={numberOfTopDownloadsToDisplay}
            />
          </section>
        </TabPanel>
        <TabPanel>
          <section className="top-pages">
            <p>
              Sessions over the last month on <strong>hostnames</strong>,
              including traffic to all web pages and app screens within that
              hostname.{" "}
              <a href={dataHrefBase + "/top-10000-domains-30-days.csv"}>
                Download the full dataset
              </a>
              .
            </p>
            <TopPagesHistorical
              dataHrefBase={dataHrefBase}
              reportFileName="top-domains-30-days.json"
              numberOfListingsToDisplay={numberOfTopPagesToDisplay}
            />
          </section>
          <section className="top-downloads">
            <div className="top-downloads__headline">
              <h3>
                Top {numberOfTopDownloadsToDisplay} Downloads Last 30 Days
              </h3>
            </div>
            <h4>
              <em>
                Total file downloads over the last month on {agency} hostnames.
              </em>
            </h4>
            <TopDownloads
              dataHrefBase={dataHrefBase}
              reportFileName="top-downloads-30-days.json"
              numberOfListingsToDisplay={numberOfTopDownloadsToDisplay}
            />
          </section>
          <section className="top-video-plays">
            <div className="top-video-plays__headline">
              <h3>
                Top {numberOfTopVideoPlaysToDisplay} Video Plays Last 30 Days
              </h3>
            </div>
            <h4>
              <em>
                Total videos played over the last month on {agency} hostnames.
              </em>
            </h4>
            <TopVideoPlays
              dataHrefBase={dataHrefBase}
              reportFileName="top-video-plays-30-days.json"
              numberOfListingsToDisplay={numberOfTopDownloadsToDisplay}
            />
          </section>
        </TabPanel>
      </Tabs>
    </section>
  );
}

SidebarContent.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  agency: PropTypes.string.isRequired,
};

export default SidebarContent;
