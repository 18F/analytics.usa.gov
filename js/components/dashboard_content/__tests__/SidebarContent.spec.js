import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import SidebarContent from "../SidebarContent";

describe("SidebarContent", () => {
  let component;

  describe("when 30 minutes tab is selected", () => {
    beforeEach(async () => {
      component = render(
        <SidebarContent
          dataHrefBase="http://www.example.com/data/"
          agency="Department of Interior"
        />,
      );
      await waitFor(() => screen.getByRole("button", { name: "30 mins" }));
    });

    it("renders a component with top pages realtime, and top downloads yesterday", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when 7 days tab is selected", () => {
    beforeEach(async () => {
      const user = userEvent.setup();

      component = render(
        <SidebarContent
          dataHrefBase="http://www.example.com/data/"
          agency="Department of Interior"
        />,
      );
      await waitFor(() => screen.getByRole("button", { name: "7 days" }));
      await user.click(screen.getByRole("button", { name: "7 days" }));
    });

    it("renders a component with top pages for 7 days, and top downloads for 7 days", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when 30 days tab is selected", () => {
    beforeEach(async () => {
      const user = userEvent.setup();

      component = render(
        <SidebarContent
          dataHrefBase="http://www.example.com/data/"
          agency="Department of Interior"
        />,
      );
      await waitFor(() => screen.getByRole("button", { name: "30 days" }));
      await user.click(screen.getByRole("button", { name: "30 days" }));
    });

    it("renders a component with top pages for 30 days, and top downloads for 30 days", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });
});
