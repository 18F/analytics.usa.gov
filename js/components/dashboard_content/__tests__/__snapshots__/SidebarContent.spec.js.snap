// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarContent when 7 days tab is selected renders a component with top pages for 7 days, and top downloads for 7 days 1`] = `
<DocumentFragment>
  <section
    class="sidebar-content"
  >
    <div
      class="sidebar-content__headline"
    >
      <h2>
        Top 30 Web Pages and App Screens
      </h2>
    </div>
    <div
      class="react-tabs"
      data-rttabs="true"
    >
      <ul
        class="usa-button-group usa-button-group--segmented"
        role="tablist"
      >
        <li
          aria-controls="panel:r1:0"
          aria-disabled="false"
          aria-selected="false"
          class="usa-button-group__item"
          data-rttab="true"
          id="tab:r1:0"
          role="tab"
        >
          <button
            class="usa-button"
            type="button"
          >
            30 mins
          </button>
        </li>
        <li
          aria-controls="panel:r1:1"
          aria-disabled="false"
          aria-selected="true"
          class="usa-button-group__item react-tabs__tab--selected"
          data-rttab="true"
          id="tab:r1:1"
          role="tab"
          tabindex="0"
        >
          <button
            class="usa-button"
            type="button"
          >
            7 days
          </button>
        </li>
        <li
          aria-controls="panel:r1:2"
          aria-disabled="false"
          aria-selected="false"
          class="usa-button-group__item"
          data-rttab="true"
          id="tab:r1:2"
          role="tab"
        >
          <button
            class="usa-button"
            type="button"
          >
            30 days
          </button>
        </li>
      </ul>
      <div
        aria-labelledby="tab:r1:0"
        class="react-tabs__tab-panel"
        id="panel:r1:0"
        role="tabpanel"
      />
      <div
        aria-labelledby="tab:r1:1"
        class="react-tabs__tab-panel react-tabs__tab-panel--selected"
        id="panel:r1:1"
        role="tabpanel"
      >
        <section
          class="top-pages"
        >
          <p>
            Sessions over the last week on 
            <strong>
              hostnames
            </strong>
            , including traffic to all web pages and app screens within that hostname.
          </p>
          <figure
            class="top-pages__bar-chart"
            data-source="http://www.example.com/data//top-domains-7-days.json"
          >
            <div
              class="data bar-chart"
            />
          </figure>
        </section>
        <section
          class="top-downloads"
        >
          <div
            class="top-downloads__headline"
          >
            <h3>
              Top 10 Downloads Last 7 Days
            </h3>
          </div>
          <h4>
            <em>
              Total file downloads over the last week on Department of Interior hostnames.
            </em>
          </h4>
          <figure
            class="top-downloads__bar-chart"
          >
            <div
              class="data bar-chart"
            />
          </figure>
        </section>
        <section
          class="top-video-plays"
        >
          <div
            class="top-video-plays__headline"
          >
            <h3>
              Top 10 Video Plays Last 7 Days
            </h3>
          </div>
          <h4>
            <em>
              Total videos played over the last week on Department of Interior hostnames.
            </em>
          </h4>
          <figure
            class="top-video-plays__bar-chart"
          >
            <div
              class="data bar-chart"
            />
          </figure>
        </section>
      </div>
      <div
        aria-labelledby="tab:r1:2"
        class="react-tabs__tab-panel"
        id="panel:r1:2"
        role="tabpanel"
      />
    </div>
  </section>
</DocumentFragment>
`;

exports[`SidebarContent when 30 days tab is selected renders a component with top pages for 30 days, and top downloads for 30 days 1`] = `
<DocumentFragment>
  <section
    class="sidebar-content"
  >
    <div
      class="sidebar-content__headline"
    >
      <h2>
        Top 30 Web Pages and App Screens
      </h2>
    </div>
    <div
      class="react-tabs"
      data-rttabs="true"
    >
      <ul
        class="usa-button-group usa-button-group--segmented"
        role="tablist"
      >
        <li
          aria-controls="panel:r2:0"
          aria-disabled="false"
          aria-selected="false"
          class="usa-button-group__item"
          data-rttab="true"
          id="tab:r2:0"
          role="tab"
        >
          <button
            class="usa-button"
            type="button"
          >
            30 mins
          </button>
        </li>
        <li
          aria-controls="panel:r2:1"
          aria-disabled="false"
          aria-selected="false"
          class="usa-button-group__item"
          data-rttab="true"
          id="tab:r2:1"
          role="tab"
        >
          <button
            class="usa-button"
            type="button"
          >
            7 days
          </button>
        </li>
        <li
          aria-controls="panel:r2:2"
          aria-disabled="false"
          aria-selected="true"
          class="usa-button-group__item react-tabs__tab--selected"
          data-rttab="true"
          id="tab:r2:2"
          role="tab"
          tabindex="0"
        >
          <button
            class="usa-button"
            type="button"
          >
            30 days
          </button>
        </li>
      </ul>
      <div
        aria-labelledby="tab:r2:0"
        class="react-tabs__tab-panel"
        id="panel:r2:0"
        role="tabpanel"
      />
      <div
        aria-labelledby="tab:r2:1"
        class="react-tabs__tab-panel"
        id="panel:r2:1"
        role="tabpanel"
      />
      <div
        aria-labelledby="tab:r2:2"
        class="react-tabs__tab-panel react-tabs__tab-panel--selected"
        id="panel:r2:2"
        role="tabpanel"
      >
        <section
          class="top-pages"
        >
          <p>
            Sessions over the last month on 
            <strong>
              hostnames
            </strong>
            , including traffic to all web pages and app screens within that hostname. 
            <a
              href="http://www.example.com/data//top-10000-domains-30-days.csv"
            >
              Download the full dataset
            </a>
            .
          </p>
          <figure
            class="top-pages__bar-chart"
            data-source="http://www.example.com/data//top-domains-30-days.json"
          >
            <div
              class="data bar-chart"
            />
          </figure>
        </section>
        <section
          class="top-downloads"
        >
          <div
            class="top-downloads__headline"
          >
            <h3>
              Top 10 Downloads Last 30 Days
            </h3>
          </div>
          <h4>
            <em>
              Total file downloads over the last month on Department of Interior hostnames.
            </em>
          </h4>
          <figure
            class="top-downloads__bar-chart"
          >
            <div
              class="data bar-chart"
            />
          </figure>
        </section>
        <section
          class="top-video-plays"
        >
          <div
            class="top-video-plays__headline"
          >
            <h3>
              Top 10 Video Plays Last 30 Days
            </h3>
          </div>
          <h4>
            <em>
              Total videos played over the last month on Department of Interior hostnames.
            </em>
          </h4>
          <figure
            class="top-video-plays__bar-chart"
          >
            <div
              class="data bar-chart"
            />
          </figure>
        </section>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`SidebarContent when 30 minutes tab is selected renders a component with top pages realtime, and top downloads yesterday 1`] = `
<DocumentFragment>
  <section
    class="sidebar-content"
  >
    <div
      class="sidebar-content__headline"
    >
      <h2>
        Top 30 Web Pages and App Screens
      </h2>
    </div>
    <div
      class="react-tabs"
      data-rttabs="true"
    >
      <ul
        class="usa-button-group usa-button-group--segmented"
        role="tablist"
      >
        <li
          aria-controls="panel:r0:0"
          aria-disabled="false"
          aria-selected="true"
          class="usa-button-group__item react-tabs__tab--selected"
          data-rttab="true"
          id="tab:r0:0"
          role="tab"
          tabindex="0"
        >
          <button
            class="usa-button"
            type="button"
          >
            30 mins
          </button>
        </li>
        <li
          aria-controls="panel:r0:1"
          aria-disabled="false"
          aria-selected="false"
          class="usa-button-group__item"
          data-rttab="true"
          id="tab:r0:1"
          role="tab"
        >
          <button
            class="usa-button"
            type="button"
          >
            7 days
          </button>
        </li>
        <li
          aria-controls="panel:r0:2"
          aria-disabled="false"
          aria-selected="false"
          class="usa-button-group__item"
          data-rttab="true"
          id="tab:r0:2"
          role="tab"
        >
          <button
            class="usa-button"
            type="button"
          >
            30 days
          </button>
        </li>
      </ul>
      <div
        aria-labelledby="tab:r0:0"
        class="react-tabs__tab-panel react-tabs__tab-panel--selected"
        id="panel:r0:0"
        role="tabpanel"
      >
        <section
          class="top-pages"
        >
          <p>
            <strong>
              Users
            </strong>
             on a 
            <strong>
              single, specific page or app screen
            </strong>
             in the last 30 minutes. Hostnames are not currently reported in real-time, so only page title and screen name information is available.
          </p>
          <figure
            class="top-pages__realtime-bar-chart"
          >
            <div
              class="data bar-chart"
            />
          </figure>
        </section>
        <section
          class="top-downloads"
        >
          <div
            class="top-downloads__headline"
          >
            <h3>
              Top 10 Downloads Yesterday
            </h3>
          </div>
          <h4>
            <em>
              Total file downloads yesterday on Department of Interior hostnames.
            </em>
          </h4>
          <figure
            class="top-downloads__bar-chart"
          >
            <div
              class="data bar-chart"
            />
          </figure>
        </section>
        <section
          class="top-video-plays"
        >
          <div
            class="top-video-plays__headline"
          >
            <h3>
              Top 10 Video Plays Yesterday
            </h3>
          </div>
          <h4>
            <em>
              Total videos played yesterday on Department of Interior hostnames.
            </em>
          </h4>
          <figure
            class="top-video-plays__bar-chart"
          >
            <div
              class="data bar-chart"
            />
          </figure>
        </section>
      </div>
      <div
        aria-labelledby="tab:r0:1"
        class="react-tabs__tab-panel"
        id="panel:r0:1"
        role="tabpanel"
      />
      <div
        aria-labelledby="tab:r0:2"
        class="react-tabs__tab-panel"
        id="panel:r0:2"
        role="tabpanel"
      />
    </div>
  </section>
</DocumentFragment>
`;
